swarmrl/value_functions/generalized_advantage_estimate.py

```diff
@@ -0,0 +1,84 @@
+"""
+Module for the expected returns value function.
+"""
+import logging
+
+import jax.numpy as np
+import numpy as onp
+
+logger = logging.getLogger(__name__)
+
+
+class GAE:
+    """
+    Class for the expected returns.
+    """
+
+    def __init__(self, gamma: float = 0.99, lambda_: float = 0.95):
+        """
+        Constructor for the generalized advantage estimate  class
+
+        Parameters
+        ----------
+        gamma : float
+                A decay factor for the values of the task each time step.
+        lambda_ : float
+                A decay factor that describes the amount of bias included in the
+                advantage calculation.
+
+        Notes
+        -----
+        See https://arxiv.org/pdf/1506.02438.pdf for more information.
+        """
+        self.gamma = gamma
+        self.lambda_ = lambda_
+
+        # Set by us to stabilize division operations.
+        self.eps = np.finfo(np.float32).eps.item()
+
+    def __call__(self, rewards: np.ndarray, values: np.ndarray):
+        """
+        Call function for the advantage.
+        Parameters
+        ----------
+        rewards : np.ndarray (n_time_steps, n_particles)
+                A numpy array of rewards to use in the calculation.
+        values : np.ndarray (n_time_steps, n_particles)
+                The prediction of the critic for the episode.
+        Returns
+        -------
+        advantages : np.ndarray (n_time_steps, n_particles)
+                Expected returns for the rewards.
+        """
+        gae = 0
+        advantages = onp.zeros_like(rewards)
+        for t in reversed(range(len(rewards))):
+            if t != len(rewards) - 1:
+                delta = rewards[t] + self.gamma * values[t + 1] - values[t]
+            else:
+                delta = rewards[t] - values[t]
+
+            gae = delta + self.gamma * self.lambda_ * gae
+
+            advantages[t] = gae
+        advantages = (advantages - np.mean(advantages)) / (
+            np.std(advantages) + self.eps
+        )
+        return advantages
+
+    def returns(self, advantages: np.ndarray, values: np.ndarray):
+        """
+        Function to compute the expected return.
+        Parameters
+        ----------
+        advantages : np.ndarray (n_time_steps, n_particles)
+                A numpy array of advantages
+        values : np.ndarray (n_time_steps, n_particles)
+                The prediction of the critic for the episode.
+        Returns
+        -------
+        expected_returns : np.ndarray (n_time_steps, n_particles)
+                Expected returns for the rewards.
+        """
+        returns = advantages + values
+        return returns
```

Comment:
why not just:

```python
return advantages + values
```
This could also not be a class method as it doesn't use `self` anywhere. Tough to argue where it should go though as it also won't be called anywhere else. I guess the `__call__` method could always return both? What would that mean for the other value functions?

Reply:
- That's definitely more readable. I'll change that 
- Returning both with the call method works perfectly. 

the problem with other value functions is that, PPO does not work using expected returns. its to noisy therefore it always uses GAE.  
The structure of GAE does not fit our structure of SGD or the expected returns. 
I could change Expected Returns and SGD such that we can use GAE in SGD.
That would make it truly modular, although it might never be used in this combination.

Reply:
I'd leave it like this and fix SGD/Expected_returns in another branch. 
Especially because it is never used in a mixture that does not work. SGD/GAE or PPO/ExpRet are never used. 
And by putting a GAE and ExpRet as default in PPO and SGD we takle that for now. 
What do you think?

Reply:
Sounds good.